### DISCUSSION TITLE
Navigation and pass parameters.

### DISCUSSION THREAD

**Discussion by Midnayt** [nalu-development/nalu#52] [Q&A] [answered]
Sorry for disturbing, but I check sample and docs and didn't find how can pass params with navigation. if it's possible, can you provide some example how to handle it.

**Comment by albyrock87** [comment, marked as answer]
You're not the first one asking, so I should probably update the docs with a specific simple example :)

So here it is in the mean time.

Let's say you have your `ContactsPageModel` and when the user presses on a contact, you want to open a `ContactDetailPageModel` for the tapped contact.

Inside `ContactsPageModel` you should have your command (here I'm using CommunityToolkit MVVM as an example).

```csharp
public Task ContactSelected(Contact contact)
{
    // Create navigation descriptor
    var navigation = Navigation.Relative()
                               .Push<ContactDetailPageModel>()
                               .WithIntent(contact);
    // Invoke Nalu's INavigationService
    return _navigationService.GoToAsync(navigation);
}
```

The method `WithIntent` is what you need to pass a typed parameter to the target view model.

Then your `ContactDetailPageModel` class should implement either (or both) `IEnteringAware<Contact>` or `IAppearingAware<Contact>` depending on where you want to load the contact information inside the page.

So let's suppose loading contact information takes a while, and you want your page to display a loader and when the page appears you start loading information.

`IEnteringAware<Contact>` would block the navigation until the contact information is ready which is what we do *not* want as we prefer a fast navigation.

So you would implement `IAppearingAware<Contact>` in this specific case.

```csharp
public partial class ContactDetailPageModel(IContactLoadingService contactLoadingService)
    : ObservableObject, IAppearingAware<Contact>
{
    public async ValueTask OnAppearingAsync(Contact contact)
    {
        this.ContactDetails = await contactLoadingService.LoadContactDetails(contact);
    }
}
```

Refer to [Navigation Events](https://nalu-development.github.io/nalu/navigation.html#navigation-events) section in the documentation for more details/use cases.

**Reply by CodingOctocat** [reply]
Is it possible to understand that `IEnteringAware<Intent>` like `IQueryAttributable`?

**Reply by albyrock87** [reply]
@CodingOctocat `IEnteringAware` may be triggered at the same time of `IQueryAttrinutable` but only in a special situation: when the page is created for the first time.

For example, let's say you have `ShellContent` page `A`.

When you push `B` it will trigger `IEnteringAware` as soon as `B` instance is created and `IAppearingAware` as soon as the underlying navigation completes.

But if you then push `C` and pop `C` again go back to `B`, then only `IAppearingAware` will be triggered on `B`.

Does this clarify?

**Reply by CodingOctocat** [reply]
Thank you! In my implementation of `ApplyQueryAttributes`, I have `if (_isInitialized) return;` preventing duplicate fetching of navigation parameters, which I should be able to simplify if I use `IEnteringAware`.

**Reply by albyrock87** [reply]
@CodingOctocat correct 💯

**Comment by Midnayt** [comment]
@albyrock87 Thanks. one more question, I saw in documentation  Navigation.Relative() and  Navigation.Absolute() - whats difference?

**Reply by albyrock87** [reply]
A relative navigation starts from the current page, while absolute starts from the root.

So if your current navigation stack is `A > B > C` and you want to go back to `A` (where `A` is the view model of the shell content):

- with relative navigation you can do `.Pop().Pop()`
- with absolute you can just `.ShellContent<A>()`

It all depends on your needs.
